### PR TITLE
Adds 3D attn_mask support to merge_masks() for Multihead Attention fast path

### LINKED ISSUE #97409: nn.MultiheadAttention breaks for mask_type=2 when fast path is enabled (e.g. with batched attn_mask)
### 🐛 Describe the bug

**TLDR:** When `nn.MultiheadAttention` is used with a batched `attn_mask` which [should be](https://pytorch.org/docs/stable/generated/torch.nn.MultiheadAttention.html) shape (N*H, L, S) (where S=L for self-attn) **_and_** fast path is enabled it crashes.

_It works as expected when fast path is **not** enabled_

### Minimal example to reproduce:

```python
import torch
import torch.nn as nn


with torch.no_grad():
    B = 8
    L = 100
    D = 512
    H = 8

    mha = nn.MultiheadAttention(D, H, batch_first=True)

    X = torch.randn(B, L, D)
    M = torch.randn(B * H, L, L) > 0

    mha.train()  # disable fast path
    out, _ = mha(X, X, X, attn_mask=M, need_weights=False)  # works
    mha.eval()  # enable fast path
    out, _ = mha(X, X, X, attn_mask=M, need_weights=False)  # crashes
```

Gives:
```
RuntimeError: For mask_type == 2 mask shape should match input shape
```

I did a little debugging in the torch code around [here](https://github.com/pytorch/pytorch/blob/fe0afc58529bdf148bfda580a60e3d778b5b8f1d/torch/nn/modules/activation.py#L1176) and it seems that when `self.merge_masks` is called the `mask_type` returned is 0. However, when `torch._native_multi_head_attention` is called with `mask_type=0` but a (N*H, L, S) `attn_mask`, it seems to ignore the `mask_type` flag (see error above).

I'm on pytorch 2.0 (conda) on M2 Max macOS 13.2.1.

### Versions

Collecting environment information...
PyTorch version: 2.0.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 13.2.1 (arm64)
GCC version: Could not collect
Clang version: 14.0.0 (clang-1400.0.29.202)
CMake version: Could not collect
Libc version: N/A

Python version: 3.9.16 (main, Jan 11 2023, 10:02:19)  [Clang 14.0.6 ] (64-bit runtime)
Python platform: macOS-13.2.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M2 Max

Versions of relevant libraries:
[pip3] lion-pytorch==0.0.4
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.23.5
[pip3] pytorch-lightning==2.0.0
[pip3] pytorch-wavelets==1.3.0
[pip3] torch==2.0.0
[pip3] torch-yin==0.1.3
[pip3] torchaudio==2.0.0
[pip3] torchmetrics==0.11.1
[pip3] xitorch==0.4.0.dev0+3327cc0
[conda] lion-pytorch              0.0.4                    pypi_0    pypi
[conda] numpy                     1.23.5           py39h1398885_0  
[conda] numpy-base                1.23.5           py39h90707a3_0  
[conda] pytorch                   2.0.0                   py3.9_0    pytorch
[conda] pytorch-lightning         2.0.0                    pypi_0    pypi
[conda] pytorch-wavelets          1.3.0                    pypi_0    pypi
[conda] torch-yin                 0.1.3                    pypi_0    pypi
[conda] torchaudio                2.0.0                  py39_cpu    pytorch
[conda] torchmetrics              0.11.1                   pypi_0    pypi
[conda] xitorch                   0.4.0.dev0+3327cc0          pypi_0    pypi

cc @ezyang @gchanan @zou3519 @jbschlosser @bhosmer @cpuhrsch @erichan1

### PR DESCRIPTION
**Pull request by LucasLLC**
Fixes #97409

Adds support for 3D attn_mask by always expanding attn_mask to 4D as per https://github.com/pytorch/pytorch/pull/98375#issuecomment-1499504721

cc @jbschlosser  @bnehoran  @cpuhrsch  @erichan1

### CODE REVIEW

File: test/test_transformers.py [hunk 1]
```diff
@@ -156,6 +155,37 @@ def test_train_with_pad_and_catch_error(self, device):
...
+    def test_multiheadattention_fastpath_attn_mask(self, device, attn_mask_dim, key_padding_mask_dim):
+        with torch.no_grad():
+            B = 2
+            L = 4
+            D = 8
+            H = 4
+
+
+            if attn_mask_dim == 2:
+                attn_mask = torch.randn(L, L).to(device) > 0
```
Comment (jbschlosser) [inline test/test_transformers.py:170]: nit: you can do `torch.randn(L, L, device=device)`. Same throughout

**Review by jbschlosser — APPROVED** [review]
Thanks so much for the fix! Looks good, a couple nits but I don't want to hold this up since the cherry-pick deadline is very soon

File: test/test_transformers.py (hunk 1 above)
Comment (LucasLLC) [reply test/test_transformers.py]: Changed all calls to the recommendation 👍🏽

File: torch/nn/modules/activation.py [hunk 2]
```diff
@@ -1258,20 +1258,27 @@ def merge_masks(self, attn_mask: Optional[Tensor], key_padding_mask: Optional[Te
...
+        if attn_mask is not None:
             # In this branch query can't be a nested tensor, so it has a shape
             batch_size, seq_len, _ = query.shape
             mask_type = 2
-            key_padding_mask_expanded = key_padding_mask.view(batch_size, 1, 1, seq_len) \
-                                                        .expand(-1, self.num_heads, -1, -1)
-            attn_mask_expanded = attn_mask.view(1, 1, seq_len, seq_len).expand(batch_size, self.num_heads, -1, -1)
-            merged_mask = attn_mask_expanded + key_padding_mask_expanded
+
+            # Always expands attn_mask to 4D
```
Comment (cpuhrsch) [inline torch/nn/modules/activation.py:1270]: Are we sure this doesn't introduce any regressions further down the stack if there's assumptions around the mask being 2D?

Comment (jbschlosser) [reply torch/nn/modules/activation.py:1270]: It's passed to `masked_softmax` downstream along with the `mask_type`, which indicates an expected shape / format. AFAICT as long as the mask type fits with the output, we're good.

Comment (cpuhrsch) [reply torch/nn/modules/activation.py:1270]: @jbschlosser - I don't think so. There's a fast path within the fast path. `masked_softmax` has another check here: https://github.com/pytorch/pytorch/blob/ae55619a2bb73ebcdc80b02a6ccd72275a9ce23e/aten/src/ATen/native/cuda/SoftMax.cu#L970

Comment (jbschlosser) [reply torch/nn/modules/activation.py:1270]: we have inception fast paths? sigh.

so this PR may introduce a perf regression for 2D masks but no correctness issues, right?

Comment (cpuhrsch) [reply torch/nn/modules/activation.py:1270]: Yes, I think so.

Comment (jbschlosser) [reply torch/nn/modules/activation.py:1270]: As discussed offline, I don't think there is a regression after all. The fast-path kernel linked above is for `mask_type ==1`, which is when only a `key_padding_mask` is set, and the fix in this PR doesn't touch that case.

Interestingly, there's a TODO to write a fast-path kernel for `mask_type == 0`, `attn_mask` only set: https://github.com/pytorch/pytorch/blob/ae55619a2bb73ebcdc80b02a6ccd72275a9ce23e/aten/src/ATen/native/cuda/SoftMax.cu#L974-L976

Comment (cpuhrsch) [reply torch/nn/modules/activation.py:1270]: Yep, you're right. My bad. Still happy I pointed this out since these conditions are complex.

**Pull request closed without merging**